src/de.hpi.swa.graal.squeak/src/de/hpi/swa/graal/squeak/nodes/process/ResumeProcessNode.java

```diff
@@ -32,20 +34,29 @@ public static ResumeProcessNode create(final CompiledCodeObject code) {
 
     public abstract void executeResume(VirtualFrame frame, PointersObject newProcess);
 
-    @Specialization(guards = "hasHigherPriority(newProcess)")
+    @Specialization(guards = {"hasHigherPriority(newProcess)", "hasSuspendedContext(newProcess)"})
     protected final void doTransferTo(final VirtualFrame frame, final PointersObject newProcess,
                     @Cached final AbstractPointersObjectWriteNode pointersWriteNode,
                     @Cached("create(code)") final GetOrCreateContextNode contextNode) {
         putToSleepNode.executePutToSleep(code.image.getActiveProcess(pointersReadNode));
         contextNode.executeGet(frame).transferTo(pointersReadNode, pointersWriteNode, newProcess);
     }
 
-    @Specialization(guards = "!hasHigherPriority(newProcess)")
+    @Specialization(guards = {"!hasHigherPriority(newProcess)", "hasSuspendedContext(newProcess)"})
     protected final void doSleep(final PointersObject newProcess) {
         putToSleepNode.executePutToSleep(newProcess);
     }
 
+    @Specialization(guards = "!hasSuspendedContext(newProcess)")
+    protected final void doFail(@SuppressWarnings("unused") final PointersObject newProcess) {
+        throw PrimitiveFailed.GENERIC_ERROR;
```

Comment:
Raising a `PrimitiveFailed` in a non-primitive node is potentially dangerous, but I think this node is always part of an `AbstractPrimitiveNode`. Nonetheless, could you provide a reference into the corresponding Slang code? I was not able to find the code location that lets a primitive fail.

Reply:
InterpreterPrimitives>>primitiveResume does
	(objectMemory isContext: (objectMemory fetchPointer: SuspendedContextIndex ofObject: proc)) ifFalse:
		[^self primitiveFail].

Reply:
Thanks! In that case, the check should be moved to `PrimResumeNode`...I'm on it.

Reply:
Here it is: https://github.com/hpi-swa/graalsqueak/commit/47bdfd8055c1c254c2aa7a380ad0f2b6f505d508